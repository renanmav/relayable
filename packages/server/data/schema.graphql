input AcceptAnswerInput {
  id: ID!
  clientMutationId: String
}

type AcceptAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

"""Answer data"""
type Answer implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  question: Question
  content: String
  upvotes: Int
  downvotes: Int
  is_accepted: Boolean
  author: User
  createdAt: String
  updatedAt: String
}

input CreateAnswerInput {
  """ID of the question"""
  question: ID!
  content: String!
  clientMutationId: String
}

type CreateAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input CreateQuestionInput {
  title: String!
  content: String!
  tags: [String]
  clientMutationId: String
}

type CreateQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input DeleteQuestionInput {
  id: ID!
  clientMutationId: String
}

type DeleteQuestionPayload {
  error: String
  clientMutationId: String
}

input EditAnswerInput {
  id: ID!
  content: String!
  clientMutationId: String
}

type EditAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input EditQuestionInput {
  id: ID!
  title: String
  content: String
  tags: [String]
  clientMutationId: String
}

type EditQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input LoginWithGithubInput {
  code: String!
  clientMutationId: String
}

type LoginWithGithubPayload {
  token: String
  user: User
  clientMutationId: String
}

type Mutation {
  LoginWithGithub(input: LoginWithGithubInput!): LoginWithGithubPayload

  """Use this mutation to create a question"""
  CreateQuestion(input: CreateQuestionInput!): CreateQuestionPayload

  """Mutation to edit the question if logged in user is the author"""
  EditQuestion(input: EditQuestionInput!): EditQuestionPayload

  """Use this mutation to delete a question if you owns it"""
  DeleteQuestion(input: DeleteQuestionInput!): DeleteQuestionPayload

  """
  Mutation to view the question
  Each user represents one view
  """
  ViewQuestion(input: ViewQuestionInput!): ViewQuestionPayload

  """
  Mutation to up/down vote a question
  
  Each user vote only once (and can change his vote)
  """
  VoteQuestion(input: VoteQuestionInput!): VoteQuestionPayload

  """Use this mutation to answer a question"""
  CreateAnswer(input: CreateAnswerInput!): CreateAnswerPayload

  """Mutation to edit the answer if logged in user is the author"""
  EditAnswer(input: EditAnswerInput!): EditAnswerPayload

  """
  Mutation to up/down vote a answer
  
  Each user vote only once (and can change his vote)
  """
  VoteAnswer(input: VoteAnswerInput!): VoteAnswerPayload

  """
  Mutation to accept the answer
  Only the owner of the question can accept the answer
  Only one answer can be accepted
  """
  AcceptAnswer(input: AcceptAnswerInput!): AcceptAnswerPayload
}

type NewQuestionPayload {
  edge(token: String!): QuestionEdge
}

type NewViewPayload {
  edge(token: String!): QuestionEdge
}

type NewVotePayload {
  edge(token: String!): QuestionEdge
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""Post interface to enforce some fields"""
interface Post {
  content: String
  upvotes: Int
  downvotes: Int
  author: User
  createdAt: String
  updatedAt: String
}

"""The root of all queries"""
type Query {
  """Use this query to fetch GitHub login URL"""
  githubLoginUrl: String

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node

  """Get your own data based on provided Authorization token"""
  me: User

  """Fetch a user by its login"""
  user(login: String!): User

  """
  Returns a connection with connectionArgs and/or the Author global ID
  
  You can search questions too with the search arg
  
  Search its done by title, content and tags fields
  """
  questions(after: String, first: Int, before: String, last: Int, authorId: ID, search: String): QuestionConnection

  """
  Returns the average difference between a question creation and its first answer
  
  Value returned is in minutes
  """
  questionAvgResponse: String
}

"""Question data"""
type Question implements Node & Post {
  """The ID of an object"""
  id: ID!
  _id: String
  title: String
  content: String
  upvotes: Int
  downvotes: Int
  views: Int
  anonymous_views: Int
  total_views: Int
  tags: [String]
  author: User
  answers: [Answer]
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type QuestionConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [QuestionEdge]
}

"""An edge in a connection."""
type QuestionEdge {
  """The item at the end of the edge"""
  node: Question

  """A cursor for use in pagination"""
  cursor: String!
}

type Subscription {
  NewQuestion: NewQuestionPayload
  NewView: NewViewPayload
  NewVote: NewVotePayload
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  github_id: String
  name: String
  login: String
  avatar_url: String
}

input ViewQuestionInput {
  id: ID!
  clientMutationId: String
}

type ViewQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}

input VoteAnswerInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteAnswerPayload {
  error: String
  answer: Answer
  clientMutationId: String
}

input VoteQuestionInput {
  id: ID!
  up: Boolean
  down: Boolean
  clientMutationId: String
}

type VoteQuestionPayload {
  error: String
  question: Question
  clientMutationId: String
}
