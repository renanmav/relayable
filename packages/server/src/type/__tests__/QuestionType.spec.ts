import { graphql } from 'graphql'
import { toGlobalId } from 'graphql-relay'

import { schema } from '../../schema'
import {
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
  getContext,
  createRows,
} from '../../../test/helper'

beforeAll(connectMongoose)

beforeEach(clearDbAndRestartCounters)

afterAll(disconnectMongoose)

const gql = String.raw

const query = gql`
  query QuestionConnection($authorId: ID, $search: String) {
    questions(authorId: $authorId, search: $search) {
      edges {
        node {
          content
          tags
          author {
            name
          }
        }
      }
    }
  }
`

const rootValue = {}

describe('when authenticated', () => {
  it('should retrieve the Question connection and its data', async () => {
    const user = await createRows.createUser()
    await createRows.createQuestion({ author: user._id })

    const context = getContext({ user })

    const result = await graphql(schema, query, rootValue, context)

    expect(result).toMatchSnapshot()
  })

  it('should retrieve only questions done by user if authorId its provided', async () => {
    const userA = await createRows.createUser()
    const userB = await createRows.createUser()

    await createRows.createQuestion({ author: userA._id })
    await createRows.createQuestion({ author: userB._id })

    const context = await getContext({ user: userB })

    const variables = {
      authorId: toGlobalId('User', userA._id),
    }

    const result = await graphql(schema, query, rootValue, context, variables)

    expect(result).toMatchSnapshot()
  })

  it('should search on authorId response', async () => {
    const userA = await createRows.createUser()
    const userB = await createRows.createUser()

    await createRows.createQuestion({ author: userA._id, tags: ['42'] })
    await createRows.createQuestion({ author: userA._id })

    await createRows.createQuestion({ author: userB._id })

    const context = await getContext({ user: userB })

    const variables = {
      authorId: toGlobalId('User', userA._id),
      search: '42',
    }

    const result = await graphql(schema, query, rootValue, context, variables)

    expect(result).toMatchSnapshot()
  })
})

describe('when unauthenticated', () => {
  it('should retrieve the Question connection', async () => {
    const user = await createRows.createUser()
    await createRows.createQuestion({ author: user._id })

    const context = getContext()

    const result = await graphql(schema, query, rootValue, context)

    expect(result).toMatchSnapshot()
  })
})

it('should return how much time a question takes to get answered', async () => {
  const user = await createRows.createUser()
  const question = await createRows.createQuestion({ author: user._id })

  const context = getContext({ user })

  const answerQuery = gql`
    mutation AnswerQuery($question: ID!) {
      CreateAnswer(input: { question: $question, content: "Something" }) {
        error
      }
    }
  `
  const variables = {
    question: toGlobalId('Question', question._id),
  }

  const answerResponse = await graphql(schema, answerQuery, rootValue, context, variables)

  expect(answerResponse).toMatchSnapshot()

  const avgResponseQuery = gql`
    query {
      questionAvgResponse
    }
  `

  const avgResponse = await graphql(schema, avgResponseQuery, rootValue, context)

  expect(avgResponse).toMatchSnapshot()
})
