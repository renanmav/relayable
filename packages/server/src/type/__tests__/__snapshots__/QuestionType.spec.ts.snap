// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return how much time a question takes to get answered 1`] = `
Object {
  "data": Object {
    "CreateAnswer": Object {
      "error": null,
    },
  },
}
`;

exports[`should return how much time a question takes to get answered 2`] = `
Object {
  "data": Object {
    "questionAvgResponse": "a few seconds",
  },
}
`;

exports[`when authenticated should retrieve only questions done by user if authorId its provided 1`] = `
Object {
  "data": Object {
    "questions": Object {
      "edges": Array [
        Object {
          "node": Object {
            "author": Object {
              "name": "User 1",
            },
            "content": "What is the meaning of life?",
            "tags": Array [],
          },
        },
      ],
    },
  },
}
`;

exports[`when authenticated should retrieve the Question connection and its data 1`] = `
Object {
  "data": Object {
    "questions": Object {
      "edges": Array [
        Object {
          "node": Object {
            "author": Object {
              "name": "User 1",
            },
            "content": "What is the meaning of life?",
            "tags": Array [],
          },
        },
      ],
    },
  },
}
`;

exports[`when authenticated should search on authorId response 1`] = `
Object {
  "data": Object {
    "questions": Object {
      "edges": Array [
        Object {
          "node": Object {
            "author": Object {
              "name": "User 1",
            },
            "content": "What is the meaning of life?",
            "tags": Array [
              "42",
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`when unauthenticated should retrieve the Question connection 1`] = `
Object {
  "data": Object {
    "questions": Object {
      "edges": Array [
        Object {
          "node": Object {
            "author": Object {
              "name": "User 1",
            },
            "content": "What is the meaning of life?",
            "tags": Array [],
          },
        },
      ],
    },
  },
}
`;
